File: AdBoard.Api/src/main/kotlin/ru/atom/adboard/dal/entities/Announcement.kt
package ru.atom.adboard.dal.entities

import com.fasterxml.jackson.annotation.JsonBackReference
import com.fasterxml.jackson.annotation.JsonIgnore
import jakarta.persistence.*
import lombok.EqualsAndHashCode
import java.util.*
import kotlin.collections.HashSet

@Entity
@Table(name = "announcements")
@EqualsAndHashCode
data class Announcement(
    @Id
    @Column(name = "id", columnDefinition = "UUID")
    val id: UUID = UUID.randomUUID(),

    @Column(name = "creator_Id", columnDefinition = "UUID", nullable = false)
    val creatorId: UUID,

    @Column(name = "title", nullable = false)
    var title: String,

    @Column(name = "description")
    var description: String,

    @Column(name = "price", nullable = false)
    var price: Double,

    @Column(name = "city")
    var city: String,

    @Column(name = "count")
    var count: Int,

    @Column(name = "subcategory_id")
    @JsonIgnore
    val subcategoryId: UUID,

    @Column(name = "is_hidden", nullable = false)
    var isHidden: Boolean = false,

    @OneToMany(mappedBy = "announcement", cascade = [CascadeType.ALL], orphanRemoval = true, fetch = FetchType.LAZY)
    @JsonIgnore
    val reviews: MutableSet<Review> = mutableSetOf(),

    @ManyToMany(mappedBy = "favorites")
    @JsonIgnore
    val favoritedBy: MutableSet<User> = HashSet(),

    @ManyToOne
    @JoinColumn(name = "subcategory_id", insertable = false, updatable = false)
    val subcategory: Subcategory? = null,

    @ManyToOne()
    @JsonIgnore
    @JoinColumn(name = "creator_Id", insertable = false, updatable = false)
    val creator: User? = null
)
{
    constructor(creatorId: UUID, title: String, description: String, price: Double, city: String, count: Int, subcategoryId: UUID)
            : this(UUID.randomUUID(), creatorId, title, description, price, city, count, subcategoryId)

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is Announcement) return false
        return id == other.id
    }

    override fun hashCode(): Int = id.hashCode()
}

File: AdBoard.API/AdBoard.DAL/Entities/Announcement.cs
using NpgsqlTypes;
using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;
using System.Text.Json.Serialization;

namespace AdBoard.DAL.Entities;

public class Announcement
{
    public Guid Id { get; set; }

    public string? City { get; set; }

    public int? Count { get; set; }

    public Guid CreatorId { get; set; }

    public string Description { get; set; } = null!;

    public bool IsHidden { get; set; }

    public double Price { get; set; }

    public Guid? SubcategoryId { get; set; }

    public string Title { get; set; } = null!;

    public DateTime? CreatedAt { get; set; }
    [JsonIgnore]
    public NpgsqlTsVector SearchVector { get; set; } = null!;
    public User Creator { get; set; } = null!;
    public ICollection<Image> Images { get; set; } = new List<Image>();
    public ICollection<User> InFavorites { get; set; } = new List<User>();

    public Subcategory Subcategory { get; set; } = null!;
}
